Xamarin.PropertyEditing/ViewModels/RatioViewModel.cs

```diff
@@ -0,0 +1,167 @@
+					return;
+
+				Value = new CommonRatio (Value.Numerator, value, Value.RatioSeparator);
+			}
+		}
+
+		public char RatioSeparator {
+			get { return Value.RatioSeparator; }
+
+			set {
+				if (Value.RatioSeparator.Equals (value))
+					return;
+
+				Value = new CommonRatio (Value.Numerator, Value.Denominator, value);
+			}
+		}
+
+		internal void ValueChanged (string stringValue, int caretPosition, int selectionLength, double incrementValue)
+		{
+			var separator = GetSeparator (stringValue);
+			SetRatioFromString (stringValue);
+
+			var newNumerator = Value.Numerator + incrementValue;
+			if (newNumerator < 0)
+				newNumerator = 0;
+
+			var newDenominator = Value.Denominator + incrementValue;
+			if (newDenominator < 1)
+				newDenominator = 1;
+
+			// Nothing selected, just workout out which value the caret is closest to, so we can increment it
+			if (selectionLength == 0) {
+				var sepLocation = GetSeparatorPosition (stringValue);
+				if (caretPosition <= sepLocation) {
+					Value = new CommonRatio (newNumerator, Value.Denominator, separator);
+				} else {
+					Value = new CommonRatio (Value.Numerator, newDenominator, separator);
+				}
+			} else {
+				// Increment both values.
+				Value = new CommonRatio (newNumerator, newDenominator, separator);
+			}
+		}
+
+		public string ValueString
+		{
+			get { return Value.StringValue; }
+			set {
+				if (Value.StringValue.Equals (value))
+					return;
+
+				SetRatioFromString (value);
+			}
+		}
+
+		public static char[] Separators = { ':', '/' };
+
+		public RatioViewModel (IPropertyInfo property, IEnumerable<IObjectEditor> editors)
+			: base (property, editors)
+		{
+		}
+
+		protected override void OnValueChanged ()
+		{
+			base.OnValueChanged ();
+			OnPropertyChanged (nameof (Numerator));
+			OnPropertyChanged (nameof (Denominator));
+			OnPropertyChanged (nameof (RatioSeparator));
+			OnPropertyChanged (nameof (ValueString));
+		}
+
+		internal void SetRatioFromString (string value)
+		{
+			var strippedValue = value.Replace (" ", string.Empty);
+			bool parsed = true;
+			var parts = strippedValue.Split (Separators);
+				double numerator = 0f;
+			if (parts.Length == 2) { // We have a potential ratio, let's make sure
+				if (!IsDouble (parts[0], out numerator))
+					parsed = false;
+				double denominator = 0f;
+				if (!String.IsNullOrEmpty (parts[1]) && !IsDouble (parts[1], out denominator))
+					parsed = false;
+
+				if (parsed) {
+					Value = new CommonRatio (numerator, denominator, GetSeparator (strippedValue));
+				}
+			} else if (parts.Length == 1) { // We have a potential whole number, let's make sure
+				if (!IsDouble (parts[0], out numerator)) {
+					parsed = false;
+				}
+				if (parsed) {
+					Value = new CommonRatio (numerator, 1, GetSeparator (strippedValue));
+				}
+			}
+
+			if (!parsed) {
+				SetError (string.Format (LocalizationResources.InvalidRatio, value));
+			}
+		}
+
+		char GetSeparator (string value)
```

Comment:
I think here and in `GetSeparatorPosition`, you can use `IndexOfAny`.

Reply:
I always forget about IndexOfAny.